contracts/0.8.25/vaults/VaultHub.sol

```diff
@@ -970,6 +1147,181 @@ contract VaultHub is PausableUntilWithRoles {
         return connection;
     }
 
+    /// @dev Shortcut for _settleObligations if there are any obligations
+    function _checkAndSettleObligations(
+        address _vault,
+        VaultRecord storage _record,
+        VaultConnection storage _connection,
+        uint256 _maxAllowedUnsettled
+    ) internal {
+        VaultObligations storage obligations = _vaultObligations(_vault);
+        if (obligations.unsettledLidoFees > 0 || obligations.redemptions > 0) {
+            _settleObligations(
+                _vault,
+                _record,
+                _connection,
+                obligations,
+                obligations.unsettledLidoFees,
+                obligations.redemptions,
+                _maxAllowedUnsettled
+            );
+        }
+    }
+
+    function _checkFeesAndSettleObligations(
+        address _vault,
+        VaultConnection storage _connection,
+        VaultRecord storage _record,
+        VaultObligations storage _obligations,
+        uint256 _reportCumulativeLidoFees
+    ) internal {
+        uint256 cumulativeSettledLidoFees = _obligations.cumulativeSettledLidoFees;
+        uint256 cumulativeLidoFees = cumulativeSettledLidoFees + _obligations.unsettledLidoFees;
+        if (_reportCumulativeLidoFees < cumulativeLidoFees) {
+            revert InvalidFees(_vault, _reportCumulativeLidoFees, cumulativeLidoFees);
+        }
+
+        _settleObligations(
+            _vault,
+            _record,
+            _connection,
+            _obligations,
+            _reportCumulativeLidoFees - cumulativeSettledLidoFees,
+            _obligations.redemptions,
+            _connection.pendingDisconnect ? 0 : MAX_UINT256 // in case of pending disconnect, force all obligations to be settled
+        );
+    }
+
+    /**
+     * @notice Calculates a settlement plan based on vault balance and obligations
+     * @param _vaultBalance The current balance of the vault
+     * @param _redemptions The redemptions to be settled
+     * @param _lidoFees The Lido fees to be settled
+     * @return valueToRebalance The ETH amount to be rebalanced for redemptions
+     * @return valueToLido The ETH amount to be sent to the Lido
+     * @return totalUnsettled The total ETH value of obligations remaining after the planned settlement
+     * @dev This is a pure calculation function with no side effects
+     */
+    function _planSettlement(
+        uint256 _vaultBalance,
+        uint256 _redemptions,
+        uint256 _lidoFees
+    ) internal pure returns (
+        uint256 valueToRebalance,
+        uint256 valueToLido,
+        uint256 totalUnsettled
+    ) {
+        valueToRebalance = Math256.min(_redemptions, _vaultBalance);
+        valueToLido = Math256.min(_lidoFees, _vaultBalance - valueToRebalance);
```

Comment:
:warning: dangerous

fee deduction MUST not incur rebalance or, even worse, bad debt